### PR TITLE
Normalize without 'sed'

### LINKED ISSUE #391: macOS `sed` fails to remove "singletons-x.x" prefix
Judging by the test output, it looks like this rule doesn't fire: https://github.com/goldfirere/singletons/blob/ea5787091689e857ea7c71a89ac2823644e493bd/tests/SingletonsTestSuiteUtils.hs#L178

Curiously, it seems like every other rule works alright.

This issue is specific to macOS `sed`.

### PR DESCRIPTION
**Pull request by int-index**
Closes #391

### CODE REVIEW

**Comment by goldfirere** [comment]
This adds a dependency on `turtle`, which is non-trivial. Given that this is only in the testsuite (and not in the library) I'm OK with this. I do suppose the explicit dependency on `turtle` is better than the implicit dependency on `sed`. I have not reviewed the code in detail.

**Comment by RyanGlScott** [comment]
Thanks again, @int-index! I presume that you tested that this code is portable by testing it on a macOS machine?

One possibly silly question: does the use of `turtle` over `sed` here have any appreciable performance impact? I only ask since some of the `Pattern` combinators used in this PR (e.g., `lowerBounded`) work directly over `[Char]` instead of `Text`, and I'm unclear if that adds up when you do a bunch of find-and-replacements over many golden files.

**Comment by int-index** [comment]
> I presume that you tested that this code is portable by testing it on a macOS machine?

Yes

> One possibly silly question: does the use of turtle over sed here have any appreciable performance impact?

I'll have to measure it, but I doubt that it will amount to anything significant.

**Comment by RyanGlScott** [comment]
I performed a quick-and-dirty timing comparison of the time it takes to run the test suite on `master` versus this PR (with `-j1` enabled, to avoid any noise that may be introduced due to parallelism). Here are the results for `master`:

<details>

```
Testsuite
  Singletons
    Nat:                  OK (1.25s)
    Empty:                OK (0.90s)                                                               
    Maybe:                OK (1.10s)                                                               
    BoxUnBox:             OK (0.92s)                                                               
    Operators:            OK (1.02s)                                                               
    HigherOrder:          OK (1.33s)                                                               
    Contains:             OK (1.04s)                                                               
    AsPattern:            OK (1.12s)                                                               
    DataValues:           OK (1.16s)                                                               
    EqInstances:          OK (0.89s)                                                               
    CaseExpressions:      OK (1.07s)                                                               
    Star:                 OK (1.50s)                                                               
    ReturnFunc:           OK (0.98s)                                                               
    Lambdas:              OK (1.38s)                                                               
    LambdasComprehensive: OK (1.09s)                                                               
    Error:                OK (0.98s)                                                               
    TopLevelPatterns:     OK (1.12s)                                                               
    LetStatements:        OK (1.41s)                                                               
    LambdaCase:           OK (1.14s)                                                               
    Sections:             OK (1.06s)                                                               
    PatternMatching:      OK (1.37s)                                                               
    Records:              OK (0.99s)                                                               
    T29:                  OK (1.13s)                                                               
    T33:                  OK (0.96s)                                                               
    T54:                  OK (1.03s)                                                               
    Classes:              OK (1.23s)                                                               
    Classes2:             OK (1.01s)                                                               
    FunDeps:              OK (1.06s)                                                               
    T78:                  OK (0.98s)                                                               
    OrdDeriving:          OK (2.14s)                                                               
    BoundedDeriving:      OK (1.12s)                                                               
    BadBoundedDeriving:   OK (0.46s)                                                               
    EnumDeriving:         OK (1.12s)                                                               
    BadEnumDeriving:      OK (0.44s)                                                               
    Fixity:               OK (0.97s)                                                               
    Undef:                OK (0.95s)                                                               
    T124:                 OK (0.95s)                                                               
    T136:                 OK (1.13s)                                                               
    T136b:                OK (0.97s)                                                               
    T153:                 OK (0.37s)                                                               
    T157:                 OK (0.37s)                                                               
    T159:                 OK (1.01s)                                                               
    T167:                 OK (1.09s)                                                               
    T145:                 OK (0.93s)                                                               
    PolyKinds:            OK (0.92s)                                                               
    PolyKindsApp:         OK (0.47s)                                                               
    T160:                 OK (1.00s)                                                               
    T163:                 OK (0.91s)                                                               
    T166:                 OK (0.46s)                                                               
    T172:                 OK (1.02s)                                                               
    T175:                 OK (0.42s)                                                               
    T176:                 OK (1.08s)
    T178:                 OK (1.18s)
    T183:                 OK (1.23s)
    T184:                 OK (1.42s)
    T187:                 OK (0.96s)
    T190:                 OK (1.20s)
    ShowDeriving:         OK (1.72s)
    EmptyShowDeriving:    OK (0.99s)
    StandaloneDeriving:   OK (1.54s)
    T197:                 OK (0.94s)
    T197b:                OK (0.94s)
    T200:                 OK (1.03s)
    T206:                 OK (0.36s)
    T209:                 OK (0.93s)
    T216:                 OK (0.95s)
    T226:                 OK (0.47s)
    T229:                 OK (0.90s)
    T249:                 OK (0.94s)
    OverloadedStrings:    OK (0.94s)
    T271:                 OK (1.22s)
    T287:                 OK (1.01s)
    TypeRepTYPE:          OK (0.67s)
    T297:                 OK (0.92s)
    T312:                 OK (1.03s)
    T313:                 OK (0.95s)
    T316:                 OK (0.98s)
    T322:                 OK (1.01s)
    NatSymbolReflexive:   OK (0.37s)
    T323:                 OK (0.37s)
    T332:                 OK (0.92s)
    T342:                 OK (0.92s)
    FunctorLikeDeriving:  OK (2.37s)
    T353:                 OK (0.95s)
    T358:                 OK (0.95s)
    T367:                 OK (0.96s)
    T371:                 OK (1.17s)
    T376:                 OK (0.97s)
  Promote
    Constructors:         OK (0.93s)
    GenDefunSymbols:      OK (0.93s)
    Newtypes:             OK (0.92s)
    Pragmas:              OK (0.41s)
    Prelude:              OK (1.06s)
    T180:                 OK (0.96s)
    T361:                 OK (0.91s)
  Database client
    Database:             OK (2.99s)
    Main:                 OK (0.75s)
  InsertionSort
    InsertionSortImp:     OK (1.22s)

All 98 tests passed (99.95s)
```
</details>

And here are the results for this PR:

<details>

```
Testsuite
  Singletons
    Nat:                  OK (1.24s)
    Empty:                OK (0.89s)                                                               
    Maybe:                OK (1.10s)                                                               
    BoxUnBox:             OK (0.91s)                                                               
    Operators:            OK (1.01s)                                                               
    HigherOrder:          OK (1.35s)                                                               
    Contains:             OK (1.03s)                                                               
    AsPattern:            OK (1.12s)                                                               
    DataValues:           OK (1.15s)                                                               
    EqInstances:          OK (0.89s)                                                               
    CaseExpressions:      OK (1.07s)                                                               
    Star:                 OK (1.52s)                                                               
    ReturnFunc:           OK (0.99s)                                                               
    Lambdas:              OK (1.39s)                                                               
    LambdasComprehensive: OK (1.08s)                                                               
    Error:                OK (0.97s)                                                               
    TopLevelPatterns:     OK (1.11s)                                                               
    LetStatements:        OK (1.38s)                                                               
    LambdaCase:           OK (1.13s)                                                               
    Sections:             OK (1.06s)                                                               
    PatternMatching:      OK (1.37s)                                                               
    Records:              OK (0.99s)                                                               
    T29:                  OK (1.11s)                                                               
    T33:                  OK (0.95s)                                                               
    T54:                  OK (1.02s)                                                               
    Classes:              OK (1.28s)                                                               
    Classes2:             OK (1.04s)                                                               
    FunDeps:              OK (1.04s)                                                               
    T78:                  OK (0.95s)                                                               
    OrdDeriving:          OK (2.13s)                                                               
    BoundedDeriving:      OK (1.10s)                                                               
    BadBoundedDeriving:   OK (0.45s)                                                               
    EnumDeriving:         OK (1.11s)                                                               
    BadEnumDeriving:      OK (0.45s)                                                               
    Fixity:               OK (0.98s)                                                               
    Undef:                OK (0.92s)                                                               
    T124:                 OK (0.96s)                                                               
    T136:                 OK (1.15s)                                                               
    T136b:                OK (0.95s)                                                               
    T153:                 OK (0.37s)                                                               
    T157:                 OK (0.37s)                                                               
    T159:                 OK (1.00s)                                                               
    T167:                 OK (1.09s)                                                               
    T145:                 OK (0.92s)                                                               
    PolyKinds:            OK (0.90s)                                                               
    PolyKindsApp:         OK (0.46s)                                                               
    T160:                 OK (0.99s)                                                               
    T163:                 OK (0.91s)                                                               
    T166:                 OK (0.46s)                                                               
    T172:                 OK (1.02s)                                                               
    T175:                 OK (0.42s)                                                               
    T176:                 OK (1.07s)
    T178:                 OK (1.17s)
    T183:                 OK (1.21s)
    T184:                 OK (1.48s)
    T187:                 OK (0.96s)
    T190:                 OK (1.19s)
    ShowDeriving:         OK (1.72s)
    EmptyShowDeriving:    OK (0.99s)
    StandaloneDeriving:   OK (1.54s)
    T197:                 OK (0.94s)
    T197b:                OK (0.93s)
    T200:                 OK (1.02s)
    T206:                 OK (0.36s)
    T209:                 OK (0.92s)
    T216:                 OK (0.91s)
    T226:                 OK (0.46s)
    T229:                 OK (0.90s)
    T249:                 OK (0.93s)
    OverloadedStrings:    OK (0.95s)
    T271:                 OK (1.19s)
    T287:                 OK (0.99s)
    TypeRepTYPE:          OK (0.66s)
    T297:                 OK (0.90s)
    T312:                 OK (1.03s)
    T313:                 OK (0.93s)
    T316:                 OK (0.96s)
    T322:                 OK (0.99s)
    NatSymbolReflexive:   OK (0.37s)
    T323:                 OK (0.37s)
    T332:                 OK (0.92s)
    T342:                 OK (0.91s)
    FunctorLikeDeriving:  OK (2.60s)
    T353:                 OK (0.96s)
    T358:                 OK (0.95s)
    T367:                 OK (0.94s)
    T371:                 OK (1.17s)
    T376:                 OK (0.95s)
  Promote
    Constructors:         OK (0.92s)
    GenDefunSymbols:      OK (0.93s)
    Newtypes:             OK (0.93s)
    Pragmas:              OK (0.41s)
    Prelude:              OK (1.07s)
    T180:                 OK (0.98s)
    T361:                 OK (0.89s)
  Database client
    Database:             OK (3.07s)
    Main:                 OK (0.74s)
  InsertionSort
    InsertionSortImp:     OK (1.19s)

All 98 tests passed (99.74s)
```
</details>

Overall, the effects of `turtle` appear to be negligible (in fact, it actually completed _faster_ with `turtle` when all the times are summed up). The most noticeable increase in time for an individual test on the `turtle` branch was the hefty `FunctorLikeDeriving` test, which became 0.23 seconds slower. However, this may just be a fluke, since running `FunctorLikeDeriving` repeatedly on the `turtle` branch yields times that are closer to 2.45 seconds, which is fairly close to the `master` time of 2.37 seconds.

In summary, I don't think `turtle` imposes any noticeable performance penalty, and making the test results portable more than makes up for any shuffling around of times, in my opinion. I'll go ahead and merge this. Thanks again!
